libparsec/crates/client/src/certif/shamir_recovery_setup.rs

```diff
@@ -0,0 +1,306 @@
+            ConnectionError::NoResponse(_) => Self::Offline,
+            // TODO: handle organization expired and user revoked here ?
+            err => Self::Internal(err.into()),
+        }
+    }
+}
+pub(super) async fn shamir_setup_create(
+    certificate_ops: &CertificateOps,
+    author_device: DeviceID,
+    author_id: UserID,
+    share_recipients: HashMap<UserID, u8>,
+    threshold: u8,
+) -> Result<CertificateBasedActionOutcome, CertifShamirError> {
+    // Loop is needed to deal with server requiring greater timestamp
+    let mut timestamp = certificate_ops.device.now();
+    loop {
+        let outcome = shamir_internals(
+            certificate_ops,
+            author_device,
+            author_id,
+            &share_recipients,
+            threshold,
+            timestamp,
+        )
+        .await?;
+
+        match outcome {
+            ShamirInternalsOutcome::Done(outcome) => return Ok(outcome),
+            ShamirInternalsOutcome::RequireGreaterTimestamp(strictly_greater_than) => {
+                // TODO: handle `strictly_greater_than` out of the client ballpark by
+                // returning an error
+                timestamp = std::cmp::max(
+                    certificate_ops.device.time_provider.now(),
+                    strictly_greater_than,
+                );
+            }
+        }
+    }
+}
+
+enum ShamirInternalsOutcome {
+    Done(CertificateBasedActionOutcome),
+    RequireGreaterTimestamp(DateTime),
+}
+
+async fn shamir_internals(
+    certificate_ops: &CertificateOps,
+    author_device: DeviceID,
+    author_id: UserID,
+    share_recipients: &HashMap<UserID, u8>,
+    threshold: u8,
+    timestamp: DateTime,
+) -> Result<ShamirInternalsOutcome, CertifShamirError> {
+    // 1. Check if share_recipients and threshold are internally coherent
+    if share_recipients
+        .iter()
+        .fold(0, |acc, (_, share_count)| acc + share_count)
+        > threshold
+    {
+        return Err(CertifShamirError::InvalidThreshold);
+    }
+
+    if share_recipients
+        .keys()
+        .find(|&&recipient| recipient == author_id)
+        .is_some()
+    {
+        return Err(CertifShamirError::AuthorIncludedAsRecipient);
+    }
+
+    if threshold == 0 {
+        return Err(CertifShamirError::ThresholdIsZero);
+    }
+
+    if share_recipients.values().find(|&&v| v == 0).is_some() {
+        return Err(CertifShamirError::ShareRecipientHasZeroShares);
+    }
+
+    // 2. Check for previous setup
+    if get_latest_shamir_setup_for_author(certificate_ops, author_device, author_id)
+        .await
+        .is_some()
+    {
+        return Err(CertifShamirError::ShamirSetupAlreadyExist);
+    }
+
+    // 3. Check organization's and author's status
+
+    // 4. Check recipients status
+
+    let mut participants_id: HashSet<_> = share_recipients.keys().collect();
+    participants_id.insert(&author_id);
+    let participants = certificate_ops.list_users(true, None, None).await?;
```

Comment:
@touilleMan I'm a bit confused about when and how to use the offset and limit args